tchannel/context.py

```diff
@@ -0,0 +1,52 @@
+# Copyright (c) 2015 Uber Technologies, Inc.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from __future__ import absolute_import
+import threading
+
+
+class TChannelLocal(threading.local):
+    __slots__ = ('context',)
+
+    def __init__(self):
+        self.context = None
+
+_LOCAL = TChannelLocal()
+_LOCAL.context = None
+
+
+class Context(object):
+    """Stack context for each TChannel instance."""
+
+    __slots__ = ('parent_tracing', '_old_context',)
+
+    def __init__(self):
+        self.parent_tracing = None
+
+    def __enter__(self):
+        self._old_context = _LOCAL.context
+        _LOCAL.context = self
+
+    def __exit__(self, type, value, traceback):
+        _LOCAL.context = self._old_context
+
+
+def get_local():
+    return _LOCAL
```

Comment:
The thread local and the stack context should not be exposed outside this module. Thread local provides one level of 'local'-ness and the stack context provides the other. These are both details that other parts of the system don't care about. They should be able to do `get_current_context` and get the 'current' `Context` object.